profiling/service/service.go

```diff
@@ -0,0 +1,147 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+// Package service defines methods to register a gRPC client/service for a
+// profiling service that is exposed in the same server. This service can be
+// queried by a client to remotely manage the gRPC profiling behaviour of an
+// application.
+package service
+
+//go:generate protoc --go_out=plugins=grpc,paths=source_relative:../proto -I../proto service.proto
+
+import (
+	"context"
+	"errors"
+	"sync"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/grpclog"
+	"google.golang.org/grpc/internal/profiling"
+	ppb "google.golang.org/grpc/profiling/proto"
+)
+
+// ProfilingConfig defines configuration options for the Init method.
+type ProfilingConfig struct {
+	// Setting this to true will enable profiling.
+	Enabled bool
+
+	// Profiling uses a circular buffer (ring buffer) to store statistics for
+	// only the last few RPCs so that profiling stats do not grow unbounded. This
+	// parameter defines the upper limit on the number of RPCs for which
+	// statistics should be stored at any given time. An average RPC requires
+	// approximately 2-3 KiB of memory for profiling-related statistics, so
+	// choose an appropriate number based on the amount of memory you can afford.
+	StreamStatsSize uint32
+
+	// To expose the profiling service and its methods, a *grpc.Server must be
+	// provided.
+	Server *grpc.Server
+}
+
+var errorNilServer = errors.New("profiling: no grpc.Server provided")
+
+// Init takes a *ProfilingConfig to initialize profiling (turned on/off
+// depending on the value set in pc.Enabled) and register the profiling service
+// in the server provided in pc.Server.
+func Init(pc *ProfilingConfig) error {
+	if pc.Server == nil {
+		return errorNilServer
+	}
+
+	if err := profiling.InitStats(pc.StreamStatsSize); err != nil {
+		return err
+	}
+
+	ppb.RegisterProfilingServer(pc.Server, getProfilingServerInstance())
+
+	// Do this last after everything has been initialized and allocated.
+	profiling.Enable(pc.Enabled)
+
+	return nil
+}
+
+type profilingServer struct {
+	drainMutex sync.Mutex
+}
+
+var profilingServerInstance *profilingServer
+var profilingServerOnce sync.Once
+
+// getProfilingServerInstance creates and returns a singleton instance of
+// profilingServer. Only one instance of profilingServer is created to use a
+// shared mutex across all profilingServer instances.
+func getProfilingServerInstance() *profilingServer {
+	profilingServerOnce.Do(func() {
+		profilingServerInstance = &profilingServer{}
+	})
+
+	return profilingServerInstance
+}
+
+func (s *profilingServer) Enable(ctx context.Context, req *ppb.EnableRequest) (*ppb.EnableResponse, error) {
+	if req.Enabled {
+		grpclog.Infof("Enabling profiling")
```

Comment:
Add a prefix to provide some extra context? `"profiling server: enabling profiling"`.  It's somewhat redundant, but "enabling profiling" is also very generic.

Reply:
Done.